### PR TITLE
Add caused_by to Diagnostic

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Hi!

Currently, miette does not support error causes which contain a `Diagnostic`. 

This draft tries to change that. It is related to #164 , as I worked with @TheNeikos on implementing this.

Tell me what you think! :heart:

---
Please note that I will rebase this PR until we all agree on a way forward!

### CODE REVIEW

File: src/handlers/graphical.rs
Comment (zkat) [inline src/handlers/graphical.rs:204]: I want `#[source]` to continue working for folks who are already using that, or who don't want to opt into `#[caused_by]`. Thoughts? It seems pretty doable.

Comment (TheNeikos) [reply src/handlers/graphical.rs:204]: IIUC, `DiagnosticChain` does do this: https://github.com/zkat/miette/pull/165/files/67b27eb94e80e27e4a4addc1a0a1bf03c4ed785f#diff-dd270622a98a63e007d36dd28857130489803df52855385b3150ef332ce4d344R61

It tries to stay on the `Diagnostic` path as long as it can

Comment (zkat) [reply src/handlers/graphical.rs:204]: No, that assumes that at the very least the toplevel "source" is a `Diagnostic`, only following the `Error::source` trail _after_ the toplevel one has been resolved. I'm talking about the case where `#[caused_by]` is never defined at all, and `thiserror`'s `#[source]` is used instead (or a manual impl of `Error::source`).

Comment (TheNeikos) [reply src/handlers/graphical.rs:204]: You're right! Totally missed that...

Comment (matthiasbeyer) [reply src/handlers/graphical.rs]: I pushed a fix which should address this.

**Comment by matthiasbeyer** [comment]
For me, tests are green now :tada:

**Comment by matthiasbeyer** [comment]
Definitively. Tell me if there's more that you want to have changed. I'll make sure to squash the history appropriately before it gets merged, if it gets merged.

File: Cargo.toml
Comment (TheNeikos) [inline Cargo.toml:3]: I think this should not be in the commit

Comment (matthiasbeyer) [reply Cargo.toml]: Whoopsi.

File: miette-derive/src/diagnostic_source.rs
Comment (zkat) [inline miette-derive/src/diagnostic_source.rs]: uhhh, this shouldn't be unused.

The `gen_struct` typically generates definitions for when the macro is used on a struct-style error, for example:

```rust
#[derive(Diagnostic, Error)]
#[error("meh")]
struct MyError {
  #[diagnostic_source]
  src: SomeError
}
```

**Comment by matthiasbeyer** [comment]
Okay, I guess now this looks like it is ready ... @zkat if you agree, I can squash and make the history here pretty before it goes in!

**Comment by zkat** [comment]
you'll want all the tests passing first :)

File: tests/test_diagnostic_source_macro.rs
Comment (zkat) [inline tests/test_diagnostic_source_macro.rs:16]: re: test failures: this needs formatting. Try `cargo fmt --all`.

File: src/handlers/graphical.rs
Comment (zkat) [inline src/handlers/graphical.rs:5]: re: test failures: this is unused now.

**Review by zkat — APPROVED** [review]
everything looks good. I'll merge once tests are passing again :)

Thanks for the contribution, and for hanging in there during the back and forth!

**Comment by matthiasbeyer** [comment]
Rebased and squashed everything no be logically grouped into commits. This includes #168 for now, but I am happy to rebase if #168 gets merged.

**Comment by zkat** [comment]
yay! Thanks again! :) Looks like we're all set~

**Pull request merged**
